pkg/inject/inject.go

```diff
@@ -0,0 +1,541 @@
+
+	if limit := conf.proxyConfig.GetResource().GetLimitCpu(); limit != "" {
+		resources.Limits["cpu"] = k8sResource.MustParse(limit)
+	}
+
+	if limit := conf.proxyConfig.GetResource().GetLimitMemory(); limit != "" {
+		resources.Limits["memory"] = k8sResource.MustParse(limit)
+	}
+
+	profileSuffixes := "."
+	if conf.proxyConfig.GetDisableExternalProfiles() {
+		profileSuffixes = "svc.cluster.local."
+	}
+	proxyUID := conf.proxyConfig.GetProxyUid()
+	sidecar := v1.Container{
+		Name:                     k8s.ProxyContainerName,
+		Image:                    conf.taggedProxyImage(),
+		ImagePullPolicy:          v1.PullPolicy(conf.proxyConfig.GetProxyImage().GetPullPolicy()),
+		TerminationMessagePolicy: v1.TerminationMessageFallbackToLogsOnError,
+		SecurityContext: &v1.SecurityContext{
+			RunAsUser: &proxyUID,
+		},
+		Ports: []v1.ContainerPort{
+			{
+				Name:          "linkerd-proxy",
+				ContainerPort: int32(conf.proxyConfig.GetInboundPort().GetPort()),
+			},
+			{
+				Name:          "linkerd-metrics",
+				ContainerPort: int32(conf.proxyConfig.GetMetricsPort().GetPort()),
+			},
+		},
+		Resources: resources,
+		Env: []v1.EnvVar{
+			{Name: "LINKERD2_PROXY_LOG", Value: conf.proxyConfig.GetLogLevel().GetLevel()},
+			{
+				Name:  "LINKERD2_PROXY_CONTROL_URL",
+				Value: fmt.Sprintf("tcp://%s:%d", controlPlaneDNS, conf.proxyConfig.GetDestinationApiPort().GetPort()),
+			},
+			{Name: "LINKERD2_PROXY_CONTROL_LISTENER", Value: fmt.Sprintf("tcp://0.0.0.0:%d", conf.proxyConfig.GetControlPort().GetPort())},
+			{Name: "LINKERD2_PROXY_METRICS_LISTENER", Value: fmt.Sprintf("tcp://0.0.0.0:%d", conf.proxyConfig.GetMetricsPort().GetPort())},
+			{Name: "LINKERD2_PROXY_OUTBOUND_LISTENER", Value: fmt.Sprintf("tcp://127.0.0.1:%d", conf.proxyConfig.GetOutboundPort().GetPort())},
+			{Name: "LINKERD2_PROXY_INBOUND_LISTENER", Value: fmt.Sprintf("tcp://0.0.0.0:%d", conf.proxyConfig.GetInboundPort().GetPort())},
+			{Name: "LINKERD2_PROXY_DESTINATION_PROFILE_SUFFIXES", Value: profileSuffixes},
+			{
+				Name:      PodNamespaceEnvVarName,
+				ValueFrom: &v1.EnvVarSource{FieldRef: &v1.ObjectFieldSelector{FieldPath: "metadata.namespace"}},
+			},
+			{Name: "LINKERD2_PROXY_INBOUND_ACCEPT_KEEPALIVE", Value: fmt.Sprintf("%dms", defaultKeepaliveMs)},
+			{Name: "LINKERD2_PROXY_OUTBOUND_CONNECT_KEEPALIVE", Value: fmt.Sprintf("%dms", defaultKeepaliveMs)},
+			{Name: "LINKERD2_PROXY_ID", Value: identity.ToDNSName()},
+		},
+		LivenessProbe:  &proxyProbe,
+		ReadinessProbe: &proxyProbe,
+	}
+
+	// Special case if the caller specifies that
+	// LINKERD2_PROXY_OUTBOUND_ROUTER_CAPACITY be set on the pod.
+	// We key off of any container image in the pod. Ideally we would instead key
+	// off of something at the top-level of the PodSpec, but there is nothing
+	// easily identifiable at that level.
+	// This is currently only used by the Prometheus pod in the control-plane.
+	for _, container := range conf.podSpec.Containers {
+		if container.Image == PrometheusImage {
```

Comment:
will this set `LINKERD2_PROXY_OUTBOUND_ROUTER_CAPACITY` into the control-plane's prometheus, or any prometheus pod matching this image name? it's probably ok if that's the case, but would be good to call out.

Reply:
Good catch. This will be set on any proxy that gets injected into any Prometheus (it is already like that in master). I'll modify the comment accordingly.

Reply:
Agreed, I had that misinterpreted. This is a nice leverage of the `resourceTransformerInject` struct :+1: